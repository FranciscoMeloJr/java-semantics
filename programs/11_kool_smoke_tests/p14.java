class a {
  int x;
  a() {
    x=7;
  }
  int m0() {
    return x;
  }
}

class b extends a {
  boolean x ;
  b() {
    super();
    x=true;
  }
  int m1() {
    return super.x;
  }
  int m2() {
    a o = this;
    return o.x;            // implicit casting
    // the field x is the one in the type a of o, that is, 7
  }
  int m3() {
    return ((a) this).x;  // explicit casting
    //"cast object to class" casts "object" to "class"
  }
  int m4() {
    return aux1(this);
  }
  int aux1(a o) {   // implicit casting, again
    return o.x;
  }
  boolean m5() {
    return aux2(this);
  }
  boolean aux2(a o) {       // implicit casting, again
    return ((b) o).x;     // but then explicit casting back to b
  }
}

class main {
  main(String[] args) {
    b o = new b();
    if (o.m5()) System.out.println(""+o.m0()+ o.m1()+ o.m2()+ o.m3()+ o.m4());
    else System.out.println("m5 not returning true");
    System.out.println("Done!");
  }
}

public class p14 {
  public static void main(String[] args) {
    new main(args);
  }
}
