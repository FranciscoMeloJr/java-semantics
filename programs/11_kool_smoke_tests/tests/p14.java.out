m5 not returning true
Done!
